### PR TITLE
chore: begin offloading computation to workers

### PR DESCRIPTION
**Pull request by harrysolovay**
[lighthouse recommendation](https://web.dev/mainthread-work-breakdown/?utm_source=lighthouse&utm_medium=devtools)

### CODE REVIEW

File: client/src/cache.worker.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {getPage as capiGetPage} from "./api/get-page";
+import {Page} from "./api/types/page";
```
Comment (sammartinez) [inline client/src/cache.worker.ts:2]: Is this file still needed since we are removing the `cache.ts` file?

File: client/src/docs-ui/internal-link/internal-link.tsx [hunk 2]
```diff
@@ -25,9 +25,9 @@ export class DocsInternalLink {
   @State() isChildActive?: boolean;
 
   @Watch("selectedFilters")
-  componentWillLoad() {
+  async componentWillLoad() {
```
Comment (sammartinez) [inline client/src/docs-ui/internal-link/internal-link.tsx:28]: Does this affect anything on Stencil side since it is there lifecycle method?

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+export const rerouteIfNecessary = async (path: string) => {
+  if (rereouteCache.has(path)) {
+    return rereouteCache.get(path);
+  }
+  const rerouted = await (async () => {
+    if (
+      path.includes("/lib") &&
+      (path.includes("/q/platform/ios") || path.includes("/q/platform/android"))
+    ) {
+      return `/lib/q/platform/${(await parseURL(path)).params
```
Comment (sammartinez) [inline client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:14]: What happens with the path param fails here? Does it break us?

File: package.json [hunk 4]
```diff
@@ -36,6 +37,7 @@
     "eslint-config-prettier": "^6.5.0",
     "eslint-plugin-prettier": "^3.1.1",
     "eslint-plugin-react": "^7.16.0",
+    "fs-extra": "^9.0.1",
```
Comment (sammartinez) [inline package.json:40]: New dependency?

File: client/src/amplify-ui/code-block/code-block.tsx [hunk 5]
```diff
@@ -59,9 +59,9 @@ export class AmplifyCodeBlock {
       return (
         <div class={lineCountStyle}>
           <div>
-              {new Array(this.lineCount)
-                .fill(null)
-                .map((_, i) => <span>{String(i + 1)}</span>)}
+            {new Array(this.lineCount).fill(null).map((_, i) => (
```
[the remark this answers is not in the document]
Comment (harrysolovay) [reply client/src/amplify-ui/code-block/code-block.tsx:62]: An update to prettier must have resulted in different formatting. I didn't write this change. Must have been the dirty work of command + s. Nbd.

File: client/src/docs-ui/internal-link/internal-link.tsx (hunk 2 above)
Comment (harrysolovay) [reply client/src/docs-ui/internal-link/internal-link.tsx:28]: Stencil's `componentWillLoad` lifecycle method can be async. It does not affect anything.

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts (hunk 3 above)
Comment (harrysolovay) [reply client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:14]: What do you mean?

File: package.json (hunk 4 above)
Comment (harrysolovay) [reply package.json:40]: Yea. It's effectively the same module as the core fs module. Just promise-based.

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+const rereouteCache = new Map<string, string>();
+
+export const rerouteIfNecessary = async (path: string) => {
+  if (rereouteCache.has(path)) {
+    return rereouteCache.get(path);
+  }
+  const rerouted = await (async () => {
+    if (
+      path.includes("/lib") &&
+      (path.includes("/q/platform/ios") || path.includes("/q/platform/android"))
```
Comment (jordanranz) [inline client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:12]: What exactly is this function doing? It seems strange to hardcode these values here when they are constants in a FILTER enum that we maintain.

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import {parseURL} from "../../utils/url/url.worker";
+
+const rereouteCache = new Map<string, string>();
+
+export const rerouteIfNecessary = async (path: string) => {
```
Comment (jordanranz) [inline client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:5]: This name is a bit ambiguous. Can we reword this?

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts (hunk 3 above)
Comment (jordanranz) [reply client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:14]: I think he is asking what happens when the underlying `serializeURL` fails

File: client/src/docs-ui/select-anchor/select-anchor.tsx [hunk 8]
```diff
@@ -62,11 +57,50 @@ export class DocsSelectAnchor {
...
+    this.selectedOption = filterKey && this.selectedFilters?.[filterKey];
+    this.selectedOptionMetadata =
+      filterKey &&
+      this.selectedOption &&
+      (filterMetadataByOptionByName[filterKey][
+        this.selectedOption
+      ] as FilterOptionMetadata);
+
+    this.options =
+      this.sortedVersions &&
```
Comment (jordanranz) [inline client/src/docs-ui/select-anchor/select-anchor.tsx:72]: Let's break this out into another function like `getOptions`

File: client/src/docs-ui/select-anchor/select-anchor.tsx [hunk 9]
```diff
@@ -62,11 +57,50 @@ export class DocsSelectAnchor {
...
+      ((await Promise.all(
+        this.sortedVersions.map(async ([filterValue, filterRoute]) => {
+          if (filterValue !== this.selectedOption) {
+            const meta =
```
Comment (jordanranz) [inline client/src/docs-ui/select-anchor/select-anchor.tsx:76]: Let's change this variable name to describe what the boolean represents

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts (hunk 3 above)
Comment (harrysolovay) [reply client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:14]: It won't (see how it ensures the path includes the virtual query params)

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts (hunk 6 above)
Comment (harrysolovay) [reply client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts:12]: Interesting. Not sure if this is needed anymore. I believe it was originally responsible for rerouting to sdk as the default. Then someone must have swapped out the "/sdk" portions of these strings with "/lib" when it went GA. Unless you can think of a reason not to, I'll get rid of this dud code.

File: client/src/docs-ui/select-anchor/select-anchor.tsx (hunk 9 above)
Comment (harrysolovay) [reply client/src/docs-ui/select-anchor/select-anchor.tsx:76]: Be more specific. What variable name?

Comment (jordanranz) [reply client/src/docs-ui/select-anchor/select-anchor.tsx:76]: `meta`

Comment (harrysolovay) [reply client/src/docs-ui/select-anchor/select-anchor.tsx]: Renamed

File: client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts (hunk 7 above)
Comment (harrysolovay) [reply client/src/docs-ui/select-anchor/reroute-if-necessary.worker.ts]: Got rid of it altogether
